File: lua/cubeconundrum.lua
local aoc = require"adventofcode"

function game_set(line)
    local set = { red=0, green=0, blue=0 }
    for num,color in line:gmatch("(%d+) +(%a+)") do
        set[color] = tonumber(num)
    end
    return set
end

function play_set(game, set)
    local function set_color(color)
        if set[color] > game[color] then
            game[color] = set[color]
        end
    end
    set_color"red"
    set_color"green"
    set_color"blue"
end

function play_game(game, line)
    local start = 1
    while start do
        local sep_start, sep_stop = line:find('; *', start)
        play_set(game, game_set(line:sub(start, sep_start)))
        start = sep_start and sep_stop+1
    end
    return game
end

function start_game(line)
    local game = { red=0, green=0, blue=0 }
    local game_num = line:match("Game (%d+): ")
    if not game_num then return nil end
    game.num = tonumber(game_num)
    line = line:sub(8 + #game_num)
    return play_game(game, line)
end

function cubeconundrum1(line, max_cubes)
    local game = start_game(line)
    if game.red > max_cubes.red
    or game.green > max_cubes.green
    or game.blue > max_cubes.blue then
        return 0
    end
    return game.num
end

function cubeconundrum2(line)
    local game = start_game(line)
    return game.red * game.green * game.blue
end

-- main
aoc.run_reduce_sum({cubeconundrum1, cubeconundrum2}, {red=12, green=13, blue=14})

File: python/cubeconundrum.py
# Advent of Code 2023
# Day 2 task: Cube Conundrum

from adventofcode import AdventOfCode
import re

GAME = re.compile("Game (\\d+): ")
CUBE = re.compile("(\\d+)\\s+(\\w+)")

class cubeconundrum1(AdventOfCode):

    def __init__(self, inputfile, maxcubes):
        self.max_cubes = maxcubes
        super().__init__(inputfile)

    def new_game(self):
        return {'red':0, 'green':0, 'blue': 0}

    def add_set(self, game, game_set):
        def game_max(color):
            game[color] = max(game[color], game_set.get(color, 0))
        game_max('red')
        game_max('green')
        game_max('blue')

    def line(self, line):
        game_line = GAME.match(line)
        game_num = int(game_line.group(1))
        game = self.new_game()
        for game_set in line[game_line.span()[1]:].split(';'):
            self.add_set(game,
                         dict( ((color,int(num)) for num,color in CUBE.findall(game_set)) )
                         )
        return self.result(game_num, game)

    def result(self, game_num, game):
        if game['red'] > self.max_cubes['red'] \
        or game['green'] > self.max_cubes['green'] \
        or game['blue'] > self.max_cubes['blue']:
            return 0
        return game_num

class cubeconundrum2(cubeconundrum1):

    def result(self, game_num, game):
        return game['red'] * game['green'] * game['blue']

def main():
    stage = cubeconundrum1
    import sys
    if len(sys.argv) > 1:
        if sys.argv[1] == '1':
            stage = cubeconundrum1
        elif sys.argv[1] == '2':
            stage = cubeconundrum2
        else:
            return
    max_cubes = {'red':12, 'green':13, 'blue':14}
    print(stage(sys.stdin, max_cubes).reduce_lines(sum))

if __name__=='__main__':
    main()
